Move pool deposits into StakePoolState

aniketd opened the pull request:
# Description

Closes #5190 

# Checklist

- [x] Commits in meaningful sequence and with useful messages.
- [ ] ~~Tests added or updated when needed.~~
- [x] `CHANGELOG.md` files updated for packages with externally visible changes.  
      **NOTE: _New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd)).
- [ ] ~~Versions updated in `.cabal` and `CHANGELOG.md` files when necessary, according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).~~
- [ ] ~~Version bounds in `.cabal` files updated when necessary.  
      **NOTE: _If bounds change in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process)).~~
- [x] Code formatted (use `scripts/fourmolize.sh`).
- [x] Cabal files formatted (use `scripts/cabal-format.sh`).
- [x] CDDL files are up to date (use `scripts/gen-cddl.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) updated (use `scripts/gen-hie.sh`).
- [x] Self-reviewed the diff.

aniketd commented on libs/cardano-ledger-core/src/Cardano/Ledger/State/CertState.hs:
```diff
@@ -494,4 +472,15 @@ psDepositsL :: Lens' (PState era) (Map (KeyHash 'StakePool) Coin)
+              Map.preserveMissing
+              (Map.zipWithMatched (\_key deposit sps -> sps {spsDeposit = deposit}))
+              deposits
+    )
```
```suggestion
```

@aniketd Remove this altogether.

aniketd replied:
In `Shelley.Rules.PoolReap` it will be better to process the reaping with the entire `StakePoolState` instead of just the deposits that are currently in `remainingDeposits`.

aniketd commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/PoolReap.hs:
```diff
@@ -125,74 +138,78 @@ instance
         (refundPools', unclaimedPools') =
           Map.partitionWithKey
-            (\cred _ -> isAccountRegistered cred accounts) -- (k ∈ dom (rewards ds))
+            (\cred _ -> isAccountRegistered cred $ cs ^. certDStateL . accountsL)
```
Keep the comment. @aniketd TODO

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/LedgerState/Types.hs:
```diff
@@ -698,11 +698,13 @@ epochStateStakePoolsL ::
+  EraCertState era => SimpleGetter (EpochState era) (Map (KeyHash 'StakePool) PoolParams)
+epochStatePoolParamsG =
+  esLStateL . lsCertStateL . certPStateL . psStakePoolsL . to (mapWithKey stakePoolStateToPoolParams)
+{-# DEPRECATED epochStatePoolParamsG "In favor of `epochStateStakePoolsL`" #-}
```
This lens is deprecated, there is no point in creating a different lens that is also deprecated. Just get rid of it
```suggestion
```

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/PoolReap.hs:
```diff
@@ -125,74 +138,78 @@ instance
-
-    ds = cs ^. certDStateL
+    -- First, activate the future stake pools.
+    cs0 = prs ^. prCertStateL
```
Switching to lenses here is bloating this PR for no reason. As I was suggesting on the call, it is better to keep various unrelated improvements to a separate PRs or at the very least to separate commits.
Regardless of this comment, using lenses here does not buy us anything.

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/PoolReap.hs:
```diff
@@ -125,74 +138,78 @@ instance
-    -- collect all of the potential refunds
-    accountRefunds :: Map.Map (Credential 'Staking) (CompactForm Coin)
-    accountRefunds =
+    retiringKeyHashes :: Set (KeyHash 'StakePool)
```
Please revert all of the unnecessary renames and "improvements" and focus only on the changes that are needed for this PR.
I've recently renamed all the bindings, no need to rename and reformat them again.

aniketd committed:
Update test examples to handle deposits correctly.

- Replace newPool/reregPool by a singular regPool.
- Remove addPoolDeposits.
- Update test examples to use this new way of handling pool deposits.
- Update ledger state golden test.

aniketd committed:
Move pool deposits from PState into StakePoolState.

* Add `spsDeposit` field to `StakePoolState`
* Remove `psDeposits` field from `PState` data constructor
* Update `mkStakePoolState` to take deposit parameter as first argument
* Replace `psDepositsL` and `psDepositsCompactL` lenses with `psDepositsG` and `psDepositsCompactG` getters
* Remove `payPoolDeposit` and `refundPoolDeposit` functions
* Update `EncCBOR`/`DecCBOR` instances for `PState` to handle new structure
* Add lenses for `StakePoolState` fields

aniketd committed:
Update shelley to use StakePoolState deposit.

Also,
* Use `mempty` deposits for genesis stake pools per specification.
* Remove `epochStatePoolParamsL` lens.
* Add prUTxOStateL, prChainAccountStateL, prCertStateL lenses to PoolReap exports.

aniketd committed:
Update shelley tests.

aniketd committed:
Update ledger-test and conformance tests.

lehins commented on eras/shelley/test-suite/src/Test/Cardano/Ledger/Shelley/Rules/AdaPreservation.hs:
```diff
@@ -236,8 +236,8 @@ checkPreservation SourceSignalTarget {source, target, signal} count =
     newRetire = lsNew ^. lsCertStateL . certPStateL . psRetiringL
-    oldPoolDeposit = lsOld ^. lsCertStateL . certPStateL . psDepositsL
-    newPoolDeposit = lsNew ^. lsCertStateL . certPStateL . psDepositsL
+    oldPoolDeposit = spsDeposit <$> lsOld ^. lsCertStateL . certPStateL . psStakePoolsL
+    newPoolDeposit = spsDeposit <$> lsNew ^. lsCertStateL . certPStateL . psStakePoolsL
```
Using Functor instance for `Map` could lead to space leaks, since it is done lazily. It is better to use `Data.Map.Strict.map` instead.
Considering that this is in the test suite it is not a big deal. But this is a general suggestion that is good to know.

aniketd replied:
Noted for the future. Thanks! :raised_hands:

lehins reviewed (approved):
Very nice! You got it almost right on point. There were a few minor things that I've noticed that could have been improved, so instead of adding those as review comments I've just created a follow up PR: #5254
